Refactor ops-traefik

mrnicegyu11 opened the pull request:
## What do these changes do?
- Make traefik configurable with env-vars instead of CLI arguments --> Better interoperability with portainer
- Remove per-cluster overwrite docker-compose files
- Remove unused env-vars and python script
## Related issue/s

## Related PR/s
https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/merge_requests/1389/
## Checklist
- [ ] I tested and it works

<!--  Extra checks based on use case -->

<!-- New Stack Introduction
- [ ] The Stack has been included in CI Workflow
-->

<!-- New Service Introduction
- [ ] Service has resource limits and reservations
- [ ] Service has placement constraints or is global
- [ ] Service is restartable
- [ ] Service restart is zero-downtime
- [ ] Service is monitored (via prometheus and grafana)
- [ ] Service is not bound to one specific node (e.g. via files or volumes)
- [ ] Relevant OPS E2E Test are added
- [ ] Service's Public URL is included in maintenance mode
- [ ] Service's Public URL is included in testing mode -->

mrnicegyu11 committed:
Refactor ops-traefik - Remove custom docker compose overwrites, add config via env-vars

mrnicegyu11 committed:
remove ununsed env-vars

mrnicegyu11 committed:
rename log-level env-vars

YuryHrytsuk commented on services/traefik/docker-compose.local.yml:
```diff
@@ -9,7 +9,7 @@ services:
       - "--api=true"
       - "--api.dashboard=true"
-      - "--log.level=${OPS_TRAEFIK_LOGLEVEL}"
+      - "--log.level=${TRAEFIK_LOG_LEVEL}"
```
Q: why removing the prefix? Since we have 2 traefiks, it makes sense, oder?

mrnicegyu11 replied:
will revert, yes

YuryHrytsuk commented on services/traefik/docker-compose.aws.yml:
```diff
@@ -1,54 +0,0 @@
-      - "--entryPoints.https.transport.respondingTimeouts.writeTimeout=21600s" #6h, for https://github.com/traefik/traefik/issues/10805
-      - "--entryPoints.https.transport.respondingTimeouts.readTimeout=21600s" #6h, for https://github.com/traefik/traefik/issues/10805
-      - "--entryPoints.smtp.address=:25"
-      - "--entryPoints.rabbit.address=:5672"
```
We need to have this port open on AWS Deployments. Otherwise, rabbit will not be reachable outside docker swarm cluster (e.g. for computational clusters logs)

mrnicegyu11 replied:
good catch thx

YuryHrytsuk commented on services/traefik/docker-compose.master.yml:
```diff
@@ -1,41 +0,0 @@
-            - "--entryPoints.https.transport.respondingTimeouts.idleTimeout=21600s" #6h, for https://github.com/traefik/traefik/issues/10805
-            - "--entryPoints.https.transport.respondingTimeouts.writeTimeout=21600s" #6h, for https://github.com/traefik/traefik/issues/10805
-            - "--entryPoints.https.transport.respondingTimeouts.readTimeout=21600s" #6h, for https://github.com/traefik/traefik/issues/10805
-            - "--entryPoints.postgres.address=:5432"
```
We need this port to be open, so that services running on Kubernetes have access to database (e.g. simcore adminer)

YuryHrytsuk replied:
We also do not want to open this port on all deployments (security)

mrnicegyu11 replied:
The entrypoint is still present, as an env-var `TRAEFIK_ENTRYPOINTS_POSTGRES_ADDRESS=":5432"` ;) just refatored

YuryHrytsuk commented on services/traefik/docker-compose.yml.j2:
```diff
@@ -183,12 +147,15 @@ services:
 
 {% endfor %}
-
+    env_file:
+        - .env
```
Q:

Is this change is needed? If so why?

mrnicegyu11 replied:
it is explicitly marking down a default, so minor

mrnicegyu11 replied:
not strictly needed, but it is best practices to make defaults explicit

YuryHrytsuk commented on services/traefik/template.env:
```diff
@@ -43,3 +38,58 @@ WEBSERVER_PORT=${WEBSERVER_PORT}
 TRAEFIK_DOMAINS_REDIRECT_IS_PERMANENT=${TRAEFIK_DOMAINS_REDIRECT_IS_PERMANENT}
+TRACING_OPENTELEMETRY_COLLECTOR_ENDPOINT=${TRACING_OPENTELEMETRY_COLLECTOR_ENDPOINT}
+TRACING_OPENTELEMETRY_COLLECTOR_PORT=${TRACING_OPENTELEMETRY_COLLECTOR_PORT}
+OPS_TRAEFIK_ACCESSLOG_ENABLED=${OPS_TRAEFIK_ACCESSLOG_ENABLED}
```
We have `TRAEFIK_LOG_LEVEL` above but `OPS_TRAEFIK_ACCESSLOG_ENABLED` here. Let's stay consistent (either with or without OPS prefix). I would for with prefix

mrnicegyu11 replied:
alright I can change this,good point

YuryHrytsuk reviewed (commented):
Thanks.

As of now we need docker-compose.master.yml (and Co) to open some ports on a subset of deployments (service exposure) while keeping this services private on other deployments where we do not need it (security)

mrnicegyu11 committed:
Merge remote-tracking branch 'upstream/main' into 2025/refactor/traefik

mrnicegyu11 committed:
only bind ports when needed

mrnicegyu11 commented:
Now using jinja2 logic and env-vars, ports are only opened when needed. The DNS 8.8.8.8 is set everywhere.

mrnicegyu11 committed:
Add env_file back to traefik compose spec

mrnicegyu11 commented on services/traefik/docker-compose.yml.j2:
```diff
@@ -2,66 +2,35 @@ version: "3.7"
       - source: traefik_dynamic_config.yml
         target: /etc/traefik/dynamic_conf.yml
+    env_file:
+      - .env
```
this is needed to so that the shell inside the container has access to the env-vars defined in .env and then passes it on to the traefik binary. The same approach is used in e.g. graylog

YuryHrytsuk replied:
It is clear what `env_file` does. 

Why do we now need env inside traefik? What change causes it?

mrnicegyu11 committed:
Merge remote-tracking branch 'upstream/main' into 2025/add/fluentd

mrnicegyu11 committed:
wip

mrnicegyu11 committed:
Merge remote-tracking branch 'upstream/main' into 2025/refactor/traefik

mrnicegyu11 committed:
Merge remote-tracking branch 'upstream/main' into 2025/refactor/traefik

YuryHrytsuk commented on services/traefik/docker-compose.yml.j2:
```diff
@@ -79,10 +48,11 @@ services:
           cpus: '1.000'
       placement:
         constraints:
+          - node.labels.traefik==true
```
Any change that makes it now necessary? This label is already present on all manager nodes?

We were moving towards direction of less labels. This is a reverse direction, I think

https://github.com/ITISFoundation/osparc-ops-environments/issues/936

mrnicegyu11 replied:
due to the removal of the `.master.` `.dalco.` etc. compose files, the constraints now have to be placed in `yml.j2`. The ops- traefik instances need to run on machines with exposed ports, managers, that on aws are connected to the ELB.

I am free to switch to the `OPS` label or something, but I think the `traefik` label is still necessary so that traefik only runs on machines that are swarm managers

YuryHrytsuk replied:
Aha, makes sense :+1:

YuryHrytsuk commented on services/traefik/template.env:
```diff
@@ -43,3 +41,61 @@ WEBSERVER_PORT=${WEBSERVER_PORT}
+TRAEFIK_ENTRYPOINTS_HTTP_TRANSPORT_RESPONDINGTIMEOUTS_IDLETIMEOUT="21600s"
+TRAEFIK_ENTRYPOINTS_HTTP_TRANSPORT_RESPONDINGTIMEOUTS_WRITETIMEOUT="21600s"
+TRAEFIK_ENTRYPOINTS_HTTP_TRANSPORT_RESPONDINGTIMEOUTS_READTIMEOUT="21600s"
+TRAEFIK_ENTRYPOINTS_RABBIT_ADDRESS=":5672"
```
avoid hardcode, use env from repo config

mrnicegyu11 replied:
since this was already hardcoded before, I would prefer to leave it as-is in this PR, not to further complicate things ;)

YuryHrytsuk commented on services/traefik/template.env:
```diff
@@ -43,3 +41,61 @@ WEBSERVER_PORT=${WEBSERVER_PORT}
+TRAEFIK_ENTRYPOINTS_HTTP_TRANSPORT_RESPONDINGTIMEOUTS_IDLETIMEOUT="21600s"
+TRAEFIK_ENTRYPOINTS_HTTP_TRANSPORT_RESPONDINGTIMEOUTS_WRITETIMEOUT="21600s"
+TRAEFIK_ENTRYPOINTS_HTTP_TRANSPORT_RESPONDINGTIMEOUTS_READTIMEOUT="21600s"
+TRAEFIK_ENTRYPOINTS_RABBIT_ADDRESS=":5672"
```
where is redis port :thinking:

mrnicegyu11 replied:
Does redis need to be exposed?

YuryHrytsuk replied:
Yes

https://github.com/ITISFoundation/osparc-ops-environments/pull/1054

YuryHrytsuk commented on services/traefik/docker-compose.yml.j2:
```diff
@@ -2,66 +2,35 @@ version: "3.7"
-        published: 5433
+      {% endif %}
+      {% if OPS_TRAEFIK_EXPOSE_RABBITMQ|lower == "true" %}
+      - target: 5672
+        published: 5672
```
avoid hardcode, use env from repo config

mrnicegyu11 replied:
It was already hardcoded before this refactoring. While you are right, I would prefer not to tie further PRs to this, it is already hanging for a while ....

YuryHrytsuk replied:
Ok.

It was not hardcoded everywhere, so I got confused.

For internal master we used env https://github.com/ITISFoundation/osparc-ops-environments/pull/1054

Fine for now

YuryHrytsuk commented on services/traefik/docker-compose.yml.j2:
```diff
@@ -2,66 +2,35 @@ version: "3.7"
-      - target: 5433
-        published: 5433
+      {% endif %}
+      {% if OPS_TRAEFIK_EXPOSE_RABBITMQ|lower == "true" %}
```
Where is redis port?

mrnicegyu11 replied:
Does redis need to be exposed?

YuryHrytsuk replied:
Yes. See https://github.com/ITISFoundation/osparc-ops-environments/pull/1054

YuryHrytsuk commented on services/traefik/docker-compose.yml.j2:
```diff
@@ -2,66 +2,35 @@ version: "3.7"
 services:
   traefik:
-    image: "traefik:v3.1.2@sha256:ec1a82940b8e00eaeef33fb4113aa1d1573b2ebb6440e10c023743fe96f08475"
+    image: "traefik:v3.4.0"
```
shall we also raise CPU Limit? @sanderegg suggested it can bring more RPS

mrnicegyu11 replied:
I would advocate to completely remove the CPU limit actually :man_shrugging:

YuryHrytsuk replied:
This is already in. Make sure to sync with current upstream

https://github.com/ITISFoundation/osparc-ops-environments/blob/main/services/traefik/docker-compose.yml.j2#L3

mrnicegyu11 replied:
true thx

YuryHrytsuk reviewed (approved):
Thanks! 

Left some comments

YuryHrytsuk commented on charts/simcore-charts/resource-usage-tracker/values.yaml.gotmpl:
```diff
@@ -2,7 +2,7 @@
 # Declare variables to be passed into your templates.
 
-replicaCount: 1
+replicaCount: {{ requiredEnv "RESOURCE_USAGE_TRACKER_KUBERNETES_REPLICAS" }}
```
Looks like this fork is behind the upstream :thinking:

mrnicegyu11 replied:
fixed, thx

mrnicegyu11 committed:
Merge remote-tracking branch 'upstream/main' into 2025/add/fluentd

mrnicegyu11 committed:
Merge remote-tracking branch 'upstream/main' into 2025/add/fluentd

mrnicegyu11 committed:
Merge branch '2025/add/fluentd' into 2025/refactor/traefik

mrnicegyu11 committed:
Merge branch '2025/add/traefikOpenTelemetry' into 2025/refactor/traefik

YuryHrytsuk committed:
Kubernetes: add local storage (#1100)

* Update longhorn README

Document how to perform (kubernetes) node maintenance

* Update Longhorn README: disks config and maintenance

* Kubernets add local storage

Use topolvm as the most mature local storage csi.

* Update longhorn readme

mrnicegyu11 committed:
Merge branch '2025/add/traefikOpenTelemetry' into 2025/refactor/traefik

mrnicegyu11 commented:
@YuryHrytsuk 

re:
![image](https://github.com/user-attachments/assets/f78c1299-4365-4702-adb3-f6a4b2f37091)


In order to remove duplication of hardcoded values, in a first step I have moved the configuration that used to be passed to traefik as CLI arguments into env-vars. This is supported by traefik according to https://doc.traefik.io/traefik/reference/static-configuration/env/

YuryHrytsuk commented on services/logging/docker-compose.yml.j2:
```diff
@@ -0,0 +1,198 @@
+version: "3.8"
```
I think this will not work in CI (aka on machines) as our servers have different docker compose verison. It will probably cause version mismatch error and fail in deploy_ops CI

I propose to ditch this label altogether. `version` is deprecated and is not used anymore (except when define may cause errors)

See https://docs.docker.com/reference/compose-file/version-and-name/#version-top-level-element-obsolete

YuryHrytsuk replied:
See example of no version definition in the newest `metabase` stack https://github.com/ITISFoundation/osparc-ops-environments/pull/1093. Works like a charm :rocket:

YuryHrytsuk commented on services/logging/docker-compose.yml.j2:
```diff
@@ -0,0 +1,198 @@
+        - traefik.http.middlewares.graylog_replace_regex.replacepathregex.regex=^/graylog/?(.*)$$
+        - traefik.http.middlewares.graylog_replace_regex.replacepathregex.replacement=/$${1}
+        - traefik.http.routers.graylog.middlewares=ops_whitelist_ips@swarm, ops_gzip@swarm, graylog_replace_regex
+  fluentd:
```
hm, this is refactor OPS traefik PR or introduce fluentd / loki PR :thinking:

mrnicegyu11 replied:
It is true, I made a bad merge and all was mixed :O I will make a fresh PR for traefik ,sorry @YuryHrytsuk

YuryHrytsuk reviewed (commented):
This is likely a very nice idea to use ENV instead of CLI Arguments as it is indeed easier to work with :ok_hand: 

Left a few comments. Also seems like `fluentd` / `loki` changes go accidentally in but should be done in a separate PR

mrnicegyu11 committed:
Fix deploy_ops CD step - monitoring

mrnicegyu11 committed:
Merge remote-tracking branch 'upstream/main' into 2025/refactor/traefik

mrnicegyu11 committed:
wip

mrnicegyu11 committed:
wip

mrnicegyu11 commented:
I made a bad merge and the code got mixed up between unrelated PRs (traefik refactoring, loki, fluentd, etc.) :O I will make a fresh PR for traefik and close this one @YuryHrytsuk
